Convert the `ColorPalette` component to TypeScript

kienstra opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
Convert the `ColorPalette` component to TypeScript

## Why?
As part of an [effort](https://github.com/WordPress/gutenberg/issues/35744) to convert `@wordpress/components` to TypeScript

## How?
By adding types to the `ColorPalette` component and its story.

## Testing Instructions
1. `npm run storybook:dev`
2. Verify that the `ColorPalette` component works the same as before

## Screenshots or screencast <!-- if applicable -->
![color-palette](https://user-images.githubusercontent.com/4063887/193477342-2ef0e1fe-9e06-4615-8dfb-7a95515fff4e.gif)

## Potential follow-ups

- Refactor all styles to Emotion
- Improve APIs — currently `__experimentalHasMultipleOrigins` relies on the fact that the format of `color` is correct. What if we inferred the single/multiple palette based on the format of `color` instead?

kienstra committed:
Convert the ColorPalette component to TypeScript

    There are some ts-ignore statements
    that would be good to remove.

kienstra committed:
Correct the PR number in the CHANGELOG

kienstra commented on packages/components/src/color-palette/index.tsx:
```diff
@@ -212,7 +223,7 @@ export default function ColorPalette( {
 	);
 
-	const colordColor = colord( value );
+	const colordColor = colord( value ?? '' );
```
`value` [can be undefined](https://github.com/WordPress/gutenberg/tree/6ffdde0670c687dee4b0569853ff76ef656961f1/packages/components/src/color-palette#value), which causes this error:

>Argument of type 'string | undefined' is not assignable to parameter of type 'AnyColor | Colord'.
>  Type 'undefined' is not assignable to type 'AnyColor | Colord'.ts(2345)
<img width="845" alt="Screen Shot 2022-10-02 at 4 26 54 PM" src="https://user-images.githubusercontent.com/4063887/193477013-8e49ce50-2b49-4ed2-8c7b-cd13a36778fd.png">

ciampo replied:
Makes sense for the scope of this PR.

From the recent discoveries while working on #44668 I also realized that `colord` will silently fail while parsing CSS variables and fallback to `#000000`, which means that all the checks / features based on `colord` are not really working for CSS variables. We should investigate this separately (cc @mirka @chad1008 )

ciampo replied:
Opened #44904 to track this separately

kienstra commented on packages/components/src/color-palette/index.tsx:
```diff
@@ -196,10 +207,10 @@ export default function ColorPalette( {
 	const clearColor = useCallback( () => onChange( undefined ), [ onChange ] );
 	const showMultiplePalettes =
-		__experimentalHasMultipleOrigins && colors?.length;
+		__experimentalHasMultipleOrigins && !! colors?.length;
```
The type of `colors?.length` is `number`, so this ensures both `__experimentalHasMultipleOrigins` and `!! colors?.length` have the type `boolean`

kienstra commented on packages/components/src/color-palette/index.tsx:
```diff
@@ -148,9 +157,9 @@ export function CustomColorPickerDropdown( {
-	currentValue,
-	colors = [],
-	showMultiplePalettes = false
+	currentValue?: string,
```
`value` is passed as the `currentValue` argument, and it [can be undefined](https://github.com/WordPress/gutenberg/blob/6ffdde0670c687dee4b0569853ff76ef656961f1/packages/components/src/color-palette/README.md#value)

kienstra commented on packages/components/src/color-palette/index.tsx:
```diff
@@ -287,11 +298,13 @@ export default function ColorPalette( {
 			<Component
 				clearable={ clearable }
 				clearColor={ clearColor }
+				// @ts-ignore Component can be MultiplePalettes or SinglePalette, which have different colors prop types.
```
Maybe I could have avoided this `@ts-ignore`, but I don't know how.

ciampo replied:
I've left a suggestion on how to avoid it in https://github.com/WordPress/gutenberg/pull/44632#pullrequestreview-1139113416

kienstra replied:
Thanks, I committed your suggestion in https://github.com/WordPress/gutenberg/pull/44632/commits/4a602dd3812c72f3422b286136a65dfd5d4a2bb3

kienstra commented:
Hi @ciampo, would you have time to review this?

You probably have a ton on your plate, thanks for any time you can give it.

Thanks, Marco!

kienstra commented:
Hi @ajitbohra,
Would you happen to have time to review this? You're probably really busy, so thanks for whatever you can spend on this.

Thanks, Ajit!

ciampo commented:
Hey @kienstra , I will try to review this today. I would normally be much faster, but I've been AFK last week and I'm still catching up — apologies!

In the meantime, I just wanted to thank you for submitting this PR in the first place!

ciampo commented on packages/components/src/color-palette/types.ts:
```diff
@@ -0,0 +1,83 @@
+export type MultipleColors = {
+	name: string;
+	colors: Color[];
+}[];
```
I wonder if we could improve the names for the `Color` and `MultipleColors` type, in a way that makes it more clear that they are related.

Maybe they could be called `ColorObject` and `PaletteObject` ? Thoughts?

kienstra replied:
Good idea, changed in https://github.com/WordPress/gutenberg/pull/44632/commits/fcd61ee3c78d1dd3f22017a96aa0de509ce7f5f5 to `ColorObject` and `PaletteObject`.

ciampo commented on packages/components/src/color-palette/types.ts:
```diff
@@ -0,0 +1,83 @@
+	clearColor: () => void;
+	onChange: OnColorChange;
+	value?: string;
+	actions: ( ReactElement | null ) | ( false | ReactElement );
```
We should be able to just use `ReactNode` for this:

```suggestion
	actions?: ReactNode;
```

kienstra replied:
That's much better, thanks!

Committed in https://github.com/WordPress/gutenberg/pull/44632/commits/c9f5d8a2f85a36641d557aae049ac2dbcb0ab6c0

ciampo replied:
Apologies in case it wasn't clear, but my suggestion was also to make `actions` non-required (via the `?` syntax) — could we do that?

kienstra replied:
Sorry, your suggestion was pretty clear! I'll commit that now.

kienstra replied:
Just committed in https://github.com/WordPress/gutenberg/pull/44632/commits/a07c43fbc934b2c75504c29fc81fcf1fabcd2664

ciampo commented on packages/components/src/color-palette/index.tsx:
```diff
@@ -33,13 +40,14 @@ function SinglePalette( {
 			const isSelected = value === color;
 
 			return (
+				// @ts-ignore Required className prop.
```
This TS error can be avoided by tweaking how props are declared in the `CircularOptionPicker` component (as explained in the [migration guide](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/CONTRIBUTING.md#refactoring-a-component-to-typescript))

<details>

<summary>Click to expand</summary>

```diff
diff --git a/packages/components/src/circular-option-picker/index.js b/packages/components/src/circular-option-picker/index.js
index 6b85a49220..f41813906f 100644
--- a/packages/components/src/circular-option-picker/index.js
+++ b/packages/components/src/circular-option-picker/index.js
@@ -16,13 +16,14 @@ import Button from '../button';
 import Dropdown from '../dropdown';
 import Tooltip from '../tooltip';
 
-function Option( {
-	className,
-	isSelected,
-	selectedIconProps,
-	tooltipText,
-	...additionalProps
-} ) {
+function Option( props ) {
+	const {
+		className,
+		isSelected,
+		selectedIconProps,
+		tooltipText,
+		...additionalProps
+	} = props;
 	const optionButton = (
 		<Button
 			isPressed={ isSelected }
@@ -52,12 +53,8 @@ function Option( {
 	);
 }
 
-function DropdownLinkAction( {
-	buttonProps,
-	className,
-	dropdownProps,
-	linkText,
-} ) {
+function DropdownLinkAction( props ) {
+	const { buttonProps, className, dropdownProps, linkText } = props;
 	return (
 		<Dropdown
 			className={ classnames(
@@ -80,7 +77,8 @@ function DropdownLinkAction( {
 	);
 }
 
-function ButtonAction( { className, children, ...additionalProps } ) {
+function ButtonAction( props ) {
+	const { className, children, ...additionalProps } = props;
 	return (
 		<Button
 			className={ classnames(
@@ -95,12 +93,8 @@ function ButtonAction( { className, children, ...additionalProps } ) {
 	);
 }
 
-export default function CircularOptionPicker( {
-	actions,
-	className,
-	options,
-	children,
-} ) {
+export default function CircularOptionPicker( props ) {
+	const { actions, className, options, children } = props;
 	return (
 		<div
 			className={ classnames(
diff --git a/packages/components/src/color-palette/index.tsx b/packages/components/src/color-palette/index.tsx
index 2330d8f61e..5ec239365f 100644
--- a/packages/components/src/color-palette/index.tsx
+++ b/packages/components/src/color-palette/index.tsx
@@ -88,7 +88,6 @@ function SinglePalette( {
 	}
 
 	return (
-		// @ts-ignore Required children prop.
 		<CircularOptionPicker
 			className={ className }
 			options={ colorOptions }
@@ -286,7 +285,6 @@ export default function ColorPalette( {
 		onChange,
 		value,
 		actions: !! clearable && (
-			// @ts-ignore Required className property.
 			<CircularOptionPicker.ButtonAction onClick={ clearColor }>
 				{ __( 'Clear' ) }
 			</CircularOptionPicker.ButtonAction>
```
</details>

kienstra replied:
Great idea! Thanks for the link to the [migration guide](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/CONTRIBUTING.md#refactoring-a-component-to-typescript), too. I didn't know about that.

Applied in https://github.com/WordPress/gutenberg/pull/44632/commits/54af17c89e963fec22550188d0aa78529d85a2df

ciampo reviewed (commented):
Hey @kienstra , it's me again! Sorry for the delay and thank you for your patience. I'm currently part of the squad that maintains the `@wordpress/components` package, so you've definitely tagged the right person to review this PR! Out of curiosity — how did you come across the TypeScript migration?

I've started taking a look at `ColorPalette`, and I must say that this is a tricky component to refactor (mostly due to a few debatable API choices around the single/multiple palettes).

From a quick look, there are many aspects that we need to refine, and therefore I've decided that **I will approach this PR review in multiple review passes**, and with each iteration we'll get closer and closer to being able to merge it.

Here's a first list of what I'm planning on focusing:

- first, I will try to help you to iron out all the `@ts-ignore` comments that are left in code
- after that, we'll focus on reviewing the existing types and using (if possible) the `WordPressComponentProps` type
- after that, I will help you to refactor any missing files to TypeScript (eg. styles, unit tests..)
- after that, I will make sure that the component still functions correctly (and potentially suggest any changes / new tests)
- finally, we will focus on Storybook and the component's docs (both in types.ts/Storybook, and README)

---

On my first review pass, I focused on getting rid of a few `@ts-ignore` comments, starting with the ones in the `BorderControl` component and the ones in `ColorPalette` caused by the single vs multiple colors duality.

Here's what I noticed:

- From existing code, it looks like `ColorPalette` actually expects that `colors` can be `undefined`
- The other main type mismatch between `ColorPalette` and `BorderControl` is that the `color` property is defined as `CSSProperties[ 'color' ]` (which also includes `undefined`), while that same property is defined as of type `string` in `ColorPalette`.
- The fact that the `MultipleColors` type is currently defined as an array seems to cause some issues in TypeScript
- The current logic in `ColorPalette` around `showMultiplePalettes` is a bit messy and not very robust:
  - it doesn't allow TypeScript to properly infer the correct type of `colors` when passed to `Component`
  - what if `colors` has a single palette format, but the `__experimentalHasMultipleOrigins` prop is set to `true`?

<details>
<summary>Following the observations above, I was able to make a few changes to the code that seem to make the code a bit more tidy and remove the need for a couple of `@ts-ignore` (click to expand)</summary>

- I aligned the type of the `color` property between `BorderControl` and `ColorPalette` — they are now both a `NonNullable< CSSProperties[ 'color' ] >`
- I removed the array type notation from `MultipleColors`, and added it inline where needed. This also allowed me to write the `colors` type as `( MultipleColors | Color )[]`, which seems to help TypeScript to properly infer it as an array at all times
- I marked the `colors` prop on `ColorPalette` as non-required, and added instead a few checks inside `ColorPalette` components to make sure that the palette UI is not rendered when the list of colors is empty
- I've added a `areColorsMultiplePalette` utility function, and used it to make sure that the format of `colors` is compatible with the value of the `__experimentalHasMultipleOrigins` prop. When the values are not compatible, the component doesn't render anything and outputs a warning (_normally we wouldn't introduce such a runtime change in a TypeScript refactor, but this particular change allows us to make TypeScript happy AND fixes a potential bug. We can decided at a later point if we want to split this change in a follow-up PR_).
- I refactored the logic around how single/multiple palettes are rendered. Instead of using a generic `Component`, I'm calling `MultiplePalettes` and `SinglePalette` explicitly, which allows us to typecast `colors` accordingly (this is made possible by the change from the previous point, which ensures that the format of `colors` is indeed matching the `__experimentalHasMultipleOrigins` prop)
- I also refactored some types in the `color-palette/index.tsx` file to reuse the `ColorPaletteProps` prop types, instead of re-declaring them as new types

```diff
diff --git a/packages/components/src/border-control/border-control-dropdown/component.tsx b/packages/components/src/border-control/border-control-dropdown/component.tsx
index 945d2ae9d6..bfe442f4c6 100644
--- a/packages/components/src/border-control/border-control-dropdown/component.tsx
+++ b/packages/components/src/border-control/border-control-dropdown/component.tsx
@@ -198,7 +198,6 @@ const BorderControlDropdown = (
 							/>
 						</HStack>
 					) : undefined }
-					{ /* @ts-ignore colors prop: Type 'Colors | undefined' is not assignable to type 'Color[] | MultipleColors' */ }
 					<ColorPalette
 						className={ popoverContentClassName }
 						value={ color }
diff --git a/packages/components/src/border-control/types.ts b/packages/components/src/border-control/types.ts
index 20cef1e4f6..a0330de20e 100644
--- a/packages/components/src/border-control/types.ts
+++ b/packages/components/src/border-control/types.ts
@@ -16,7 +16,7 @@ export type Border = {
 
 export type Color = {
 	name: string;
-	color: CSSProperties[ 'color' ];
+	color: NonNullable< CSSProperties[ 'color' ] >;
 };
 
 export type ColorOrigin = {
diff --git a/packages/components/src/color-palette/index.tsx b/packages/components/src/color-palette/index.tsx
index 8e72f4a56e..2330d8f61e 100644
--- a/packages/components/src/color-palette/index.tsx
+++ b/packages/components/src/color-palette/index.tsx
@@ -82,6 +82,11 @@ function SinglePalette( {
 			);
 		} );
 	}, [ colors, value, onChange, clearColor ] );
+
+	if ( colors.length === 0 ) {
+		return null;
+	}
+
 	return (
 		// @ts-ignore Required children prop.
 		<CircularOptionPicker
@@ -100,6 +105,10 @@ function MultiplePalettes( {
 	value,
 	actions,
 }: MultiplePalettesProps ) {
+	if ( colors.length === 0 ) {
+		return null;
+	}
+
 	return (
 		<VStack spacing={ 3 } className={ className }>
 			{ colors.map( ( { name, colors: colorPalette }, index ) => {
@@ -157,9 +166,9 @@ export function CustomColorPickerDropdown( {
 }
 
 export const extractColorNameFromCurrentValue = (
-	currentValue?: string,
-	colors: Color[] | MultipleColors = [],
-	showMultiplePalettes: boolean = false
+	currentValue?: ColorPaletteProps[ 'value' ],
+	colors: ColorPaletteProps[ 'colors' ] = [],
+	showMultiplePalettes: ColorPaletteProps[ '__experimentalHasMultipleOrigins' ] = false
 ) => {
 	if ( ! currentValue ) {
 		return '';
@@ -171,8 +180,9 @@ export const extractColorNameFromCurrentValue = (
 		: colord( currentValue ).toHex();
 
 	// Normalize format of `colors` to simplify the following loop
-	const colorPalettes = showMultiplePalettes
-		? ( colors as MultipleColors )
+	type normalizedPaletteObject = { colors: Color[] };
+	const colorPalettes: normalizedPaletteObject[] = showMultiplePalettes
+		? ( colors as MultipleColors[] )
 		: [ { colors: colors as Color[] } ];
 	for ( const { colors: paletteColors } of colorPalettes ) {
 		for ( const { name: colorName, color: colorValue } of paletteColors ) {
@@ -197,10 +207,18 @@ export const showTransparentBackground = ( currentValue?: string ) => {
 	return colord( currentValue ).alpha() === 0;
 };
 
+const areColorsMultiplePalette = (
+	colors: NonNullable< ColorPaletteProps[ 'colors' ] >
+): colors is MultipleColors[] => {
+	return colors.every( ( colorObj ) =>
+		Array.isArray( ( colorObj as MultipleColors ).colors )
+	);
+};
+
 export default function ColorPalette( {
 	clearable = true,
 	className,
-	colors,
+	colors = [],
 	disableCustomColors = false,
 	enableAlpha,
 	onChange,
@@ -209,9 +227,31 @@ export default function ColorPalette( {
 	__experimentalIsRenderedInSidebar = false,
 }: ColorPaletteProps ) {
 	const clearColor = useCallback( () => onChange( undefined ), [ onChange ] );
-	const showMultiplePalettes =
-		__experimentalHasMultipleOrigins && !! colors?.length;
-	const Component = showMultiplePalettes ? MultiplePalettes : SinglePalette;
+
+	const buttonLabelName = useMemo(
+		() =>
+			extractColorNameFromCurrentValue(
+				value,
+				colors,
+				__experimentalHasMultipleOrigins
+			),
+		[ value, colors, __experimentalHasMultipleOrigins ]
+	);
+
+	// Make sure that the `colors` array has a format (single/multiple) that is
+	// compatible with the `__experimentalHasMultipleOrigins` flag. This is true
+	// when __experimentalHasMultipleOrigins and areColorsMultiplePalette() are
+	// either both `true` or both `false`.
+	if (
+		colors.length > 0 &&
+		__experimentalHasMultipleOrigins !== areColorsMultiplePalette( colors )
+	) {
+		// eslint-disable-next-line no-console
+		console.warn(
+			'wp.components.ColorPalette: please specify a format for the `colors` prop that is compatible with the `__experimentalHasMultipleOrigins` prop.'
+		);
+		return null;
+	}
 
 	const renderCustomColorPicker = () => (
 		<DropdownContentWrapper paddingSize="none">
@@ -228,15 +268,6 @@ export default function ColorPalette( {
 	const valueWithoutLeadingHash = value?.startsWith( '#' )
 		? value.substring( 1 )
 		: value ?? '';
-	const buttonLabelName = useMemo(
-		() =>
-			extractColorNameFromCurrentValue(
-				value,
-				colors,
-				showMultiplePalettes
-			),
-		[ value, colors, showMultiplePalettes ]
-	);
 
 	const customColorAccessibleLabel = !! valueWithoutLeadingHash
 		? sprintf(
@@ -249,6 +280,19 @@ export default function ColorPalette( {
 		  )
 		: __( 'Custom color picker.' );
 
+	const paletteCommonProps = {
+		clearable,
+		clearColor,
+		onChange,
+		value,
+		actions: !! clearable && (
+			// @ts-ignore Required className property.
+			<CircularOptionPicker.ButtonAction onClick={ clearColor }>
+				{ __( 'Clear' ) }
+			</CircularOptionPicker.ButtonAction>
+		),
+	};
+
 	return (
 		<VStack spacing={ 3 } className={ className }>
 			{ ! disableCustomColors && (
@@ -295,24 +339,17 @@ export default function ColorPalette( {
 					) }
 				/>
 			) }
-			<Component
-				clearable={ clearable }
-				clearColor={ clearColor }
-				// @ts-ignore Component can be MultiplePalettes or SinglePalette, which have different colors prop types.
-				colors={ colors }
-				onChange={ onChange }
-				value={ value }
-				actions={
-					!! clearable && (
-						// @ts-ignore Required className property.
-						<CircularOptionPicker.ButtonAction
-							onClick={ clearColor }
-						>
-							{ __( 'Clear' ) }
-						</CircularOptionPicker.ButtonAction>
-					)
-				}
-			/>
+			{ __experimentalHasMultipleOrigins ? (
+				<MultiplePalettes
+					{ ...paletteCommonProps }
+					colors={ colors as MultipleColors[] }
+				/>
+			) : (
+				<SinglePalette
+					{ ...paletteCommonProps }
+					colors={ colors as Color[] }
+				/>
+			) }
 		</VStack>
 	);
 }
diff --git a/packages/components/src/color-palette/stories/index.tsx b/packages/components/src/color-palette/stories/index.tsx
index 4b88f093f7..559aadc783 100644
--- a/packages/components/src/color-palette/stories/index.tsx
+++ b/packages/components/src/color-palette/stories/index.tsx
@@ -51,7 +51,7 @@ export default meta;
 const Template: ComponentStory< typeof ColorPalette > = ( args ) => {
 	const firstColor =
 		( args.colors as Color[] )[ 0 ].color ||
-		( args.colors as MultipleColors )[ 0 ].colors[ 0 ].color;
+		( args.colors as MultipleColors[] )[ 0 ].colors[ 0 ].color;
 	const [ color, setColor ] = useState< string | undefined >( firstColor );
 
 	return (
diff --git a/packages/components/src/color-palette/types.ts b/packages/components/src/color-palette/types.ts
index 28169a939a..1376c1db38 100644
--- a/packages/components/src/color-palette/types.ts
+++ b/packages/components/src/color-palette/types.ts
@@ -1,19 +1,19 @@
 /**
  * External dependencies
  */
-import type { MouseEventHandler, ReactElement } from 'react';
+import type { CSSProperties, MouseEventHandler, ReactElement } from 'react';
 
 type OnColorChange = ( newColor?: string ) => void;
 
 export type Color = {
 	name: string;
-	color: string;
+	color: NonNullable< CSSProperties[ 'color' ] >;
 };
 
 export type MultipleColors = {
 	name: string;
 	colors: Color[];
-}[];
+};
 
 type PaletteProps = {
 	className?: string;
@@ -28,7 +28,7 @@ export type SinglePaletteProps = PaletteProps & {
 };
 
 export type MultiplePalettesProps = PaletteProps & {
-	colors: MultipleColors;
+	colors: MultipleColors[];
 };
 
 export type CustomColorPickerDropdownProps = {
@@ -52,8 +52,10 @@ export type ColorPaletteProps = {
 	className?: string;
 	/**
 	 * Array with the colors to be shown.
+	 *
+	 * @default []
 	 */
-	colors: MultipleColors | Color[];
+	colors?: ( MultipleColors | Color )[];
 	/**
 	 * Whether to allow custom color or not.
 	 */
```
</details>

kienstra commented:
Hi @ciampo, thanks so much for reviewing this and for your suggestions! I'll apply them today.

kienstra committed:
Apply Marco's suggestion to remove ts-ignore comments

Destructure props int constants after the function
signature, not inside it.

kienstra committed:
Replace complex type with ReactNode, thanks to Marco's suggestion

kienstra committed:
Apply Marco's suggestions for TS verbatim

https://github.com/WordPress/gutenberg/pull/44632#pullrequestreview-1139113416

kienstra committed:
Prevent an error from colors possibly being undefined

Types of property 'colors' are incompatible.
  Type 'Colors | undefined' is not assignable to type '(Color | MultipleColors)[]'.
    Type 'undefined' is not assignable to type '(Color | MultipleColors)[]'.

kienstra commented:
>Out of curiosity — how did you come across the TypeScript migration?

Haha, I'm not sure. 

I love TypeScript, and searched for TypeScript issues in Gutenberg 😄

kienstra committed:
Rename Color and MultipleColors to ColorObject and PaletteObject

kienstra committed:
Alphabetize the imports again

kienstra committed:
Remove another needless ts-ignore comment

kienstra commented:
>Following the observations above, I was able to make a few changes to the code that seem to make the code a bit more tidy and remove the need for a couple of `@ts-ignore` (click to expand)

Thanks, I committed your [diff](https://github.com/WordPress/gutenberg/pull/44632#pullrequestreview-1139113416) verbatim in https://github.com/WordPress/gutenberg/pull/44632/commits/4a602dd3812c72f3422b286136a65dfd5d4a2bb3.

kienstra commented on packages/components/src/border-control/border-control-dropdown/component.tsx:
```diff
@@ -198,20 +198,22 @@ const BorderControlDropdown = (
+							className={ popoverContentClassName }
+							value={ color }
+							onChange={ onColorChange }
+							{ ...{ colors, disableCustomColors } }
```
There was a TS error from the `colors` prop:

```sh
Types of property 'colors' are incompatible.
Type 'Colors | undefined' is not assignable to type '(Color | MultipleColors)[]'.
Type 'undefined' is not assignable to type '(Color | MultipleColors)[]'.
```

`colors` can be undefined, and `ColorPalette` [doesn't allow](https://github.com/kienstra/gutenberg/blob/6ee6987a2c392b9f154339e3a11d5f9ca0bb22e5/packages/components/src/color-palette/types.ts#L58) an undefined `colors` prop.

So I added a ternary to only render `ColorPalette` if `colors` is truthy.

kienstra replied:
Reverted this

kienstra committed:
Revert "Prevent an error from colors possibly being undefined"

This reverts commit 7fe648e169eb3a81cf0ebcb20755ec23ef189766.

kienstra committed:
Make colors allow undefined

kienstra committed:
Merge branch 'trunk' into update/color-palette-component-ts

kienstra commented:
Hi @ciampo, 
Just finished applying your suggestions. 

Thanks for all the care you put into your review. 

When you suggested code, I committed it verbatim.

ciampo commented:
Thank you! I'll have another round of review today or tomorrow

kienstra committed:
Make actions optional, which I forgot before

kienstra committed:
Merge in trunk, resolve conflict

There was a conflict in CHANGELOG.md,
but I kept both edits.

ciampo commented on packages/components/src/color-palette/stories/index.tsx:
```diff
@@ -92,10 +100,11 @@ MultipleOrigins.args = {
 	return (
 		<div
 			style={ {
+				// @ts-ignore
```
We can typecast the styles to avoid this `@ts-ignore` comment:

```diff
diff --git a/packages/components/src/color-palette/stories/index.tsx b/packages/components/src/color-palette/stories/index.tsx
index 51d165652d..63fac4c347 100644
--- a/packages/components/src/color-palette/stories/index.tsx
+++ b/packages/components/src/color-palette/stories/index.tsx
@@ -1,6 +1,7 @@
 /**
  * External dependencies
  */
+import type { CSSProperties } from 'react';
 import type { ComponentMeta, ComponentStory } from '@storybook/react';
 
 /**
@@ -103,12 +104,13 @@ MultipleOrigins.args = {
 export const CSSVariables = ( args: ColorPaletteProps ) => {
 	return (
 		<div
-			style={ {
-				// @ts-ignore
-				'--red': '#f00',
-				'--yellow': '#ff0',
-				'--blue': '#00f',
-			} }
+			style={
+				{
+					'--red': '#f00',
+					'--yellow': '#ff0',
+					'--blue': '#00f',
+				} as CSSProperties
+			}
 		>
 			<Template { ...args } />
 		</div>
```

kienstra replied:
Great idea, committed in https://github.com/WordPress/gutenberg/pull/44632/commits/ea5b1f3c8d0946022b351e569e825b1049b742ae

ciampo commented on packages/components/src/color-palette/stories/index.tsx:
```diff
@@ -6,11 +11,12 @@ import { useState } from '@wordpress/element';
+import type { ColorObject, PaletteObject, ColorPaletteProps } from '../types';
 
-const meta = {
+const meta: ComponentMeta< typeof ColorPalette > = {
```
Here's some improvements that we can make to the Storybook examples:

- Add a named export to `ColorPalette` to allow the Storybook docs plugin to extract type info
- remove unnecerssary type info (now that we have TypeScript) from the Storybook `meta` object
- Disable controls for `value` and `onChange`
- Add `onChange` as a Storybook action
- use the `ComponentStory` type for the `CSSVariables` story

```diff
diff --git a/packages/components/src/color-palette/index.tsx b/packages/components/src/color-palette/index.tsx
index a49c45e430..61f32c45b3 100644
--- a/packages/components/src/color-palette/index.tsx
+++ b/packages/components/src/color-palette/index.tsx
@@ -213,7 +213,7 @@ const areColorsMultiplePalette = (
 	);
 };
 
-export default function ColorPalette( {
+export function ColorPalette( {
 	clearable = true,
 	className,
 	colors = [],
@@ -350,3 +350,5 @@ export default function ColorPalette( {
 		</VStack>
 	);
 }
+
+export default ColorPalette;
diff --git a/packages/components/src/color-palette/stories/index.tsx b/packages/components/src/color-palette/stories/index.tsx
index 51d165652d..a68d1b3c88 100644
--- a/packages/components/src/color-palette/stories/index.tsx
+++ b/packages/components/src/color-palette/stories/index.tsx
@@ -14,32 +14,21 @@ import { useState } from '@wordpress/element';
 import ColorPalette from '..';
 import Popover from '../../popover';
 import { Provider as SlotFillProvider } from '../../slot-fill';
-import type { ColorObject, PaletteObject, ColorPaletteProps } from '../types';
+import type { ColorObject, PaletteObject } from '../types';
 
 const meta: ComponentMeta< typeof ColorPalette > = {
 	title: 'Components/ColorPalette',
 	component: ColorPalette,
 	argTypes: {
+		// Removing the control because setting this prop without changing the
+		// format of the `colors` prop can break the component.
 		__experimentalHasMultipleOrigins: {
 			control: {
 				type: null,
 			},
 		},
-		__experimentalIsRenderedInSidebar: {
-			control: {
-				type: 'boolean',
-			},
-		},
-		clearable: {
-			control: {
-				type: 'boolean',
-			},
-		},
-		disableCustomColors: {
-			control: {
-				type: 'boolean',
-			},
-		},
+		onChange: { action: 'onChange', control: { type: null } },
+		value: { control: { type: null } },
 	},
 	parameters: {
 		controls: { expanded: true },
@@ -48,7 +37,10 @@ const meta: ComponentMeta< typeof ColorPalette > = {
 };
 export default meta;
 
-const Template: ComponentStory< typeof ColorPalette > = ( args ) => {
+const Template: ComponentStory< typeof ColorPalette > = ( {
+	onChange,
+	...args
+} ) => {
 	const firstColor =
 		( args.colors as ColorObject[] )[ 0 ].color ||
 		( args.colors as PaletteObject[] )[ 0 ].colors[ 0 ].color;
@@ -56,7 +48,14 @@ const Template: ComponentStory< typeof ColorPalette > = ( args ) => {
 
 	return (
 		<SlotFillProvider>
-			<ColorPalette { ...args } value={ color } onChange={ setColor } />
+			<ColorPalette
+				{ ...args }
+				value={ color }
+				onChange={ ( newColor ) => {
+					setColor( newColor );
+					onChange?.( newColor );
+				} }
+			/>
 			{ /* @ts-ignore Property 'Slot' does not exist on type. */ }
 			<Popover.Slot />
 		</SlotFillProvider>
@@ -100,7 +99,7 @@ MultipleOrigins.args = {
 	],
 };
 
-export const CSSVariables = ( args: ColorPaletteProps ) => {
+export const CSSVariables: ComponentStory< typeof ColorPalette > = ( args ) => {
 	return (
 		<div
 			style={ {
```

kienstra replied:
Good ideas, committed in https://github.com/WordPress/gutenberg/pull/44632/commits/c4202328e9cca48c41fb7d32e13bf9fb3e370771

ciampo commented on packages/components/src/color-palette/types.ts:
```diff
@@ -0,0 +1,85 @@
+	/**
+	 * Whether the palette should have a clearing button.
+	 */
+	clearable?: boolean;
```
Looking at the [component's props destructuring](https://github.com/WordPress/gutenberg/blob/6ee6987a2c392b9f154339e3a11d5f9ca0bb22e5/packages/components/src/color-palette/index.tsx#L217-L225), a few props are missing their `@default` type annotation

```diff
diff --git a/packages/components/src/color-palette/types.ts b/packages/components/src/color-palette/types.ts
index 1c8f6ec1a8..a940ff09a0 100644
--- a/packages/components/src/color-palette/types.ts
+++ b/packages/components/src/color-palette/types.ts
@@ -44,6 +44,8 @@ export type CustomColorPickerDropdownProps = {
 export type ColorPaletteProps = {
 	/**
 	 * Whether the palette should have a clearing button.
+	 *
+	 * @default true
 	 */
 	clearable?: boolean;
 	/**
@@ -58,6 +60,8 @@ export type ColorPaletteProps = {
 	colors?: ( PaletteObject | ColorObject )[];
 	/**
 	 * Whether to allow custom color or not.
+	 *
+	 * @default false
 	 */
 	disableCustomColors?: boolean;
 	/**
@@ -76,10 +80,14 @@ export type ColorPaletteProps = {
 	/**
 	 * Whether the colors prop is an array of color palettes,
 	 * rather than an array of color objects.
+	 *
+	 * @default false
 	 */
 	__experimentalHasMultipleOrigins?: boolean;
 	/**
 	 * Whether this is rendered in the sidebar.
+	 *
+	 * @default false
 	 */
 	__experimentalIsRenderedInSidebar?: boolean;
 };
```

kienstra replied:
Thanks, committed in https://github.com/WordPress/gutenberg/pull/44632/commits/9b770d3b07dd216aafb9c566deadd15ea5842794

ciampo commented on packages/components/src/color-palette/index.tsx:
```diff
@@ -179,28 +198,58 @@ export const extractColorNameFromCurrentValue = (
 	__experimentalHasMultipleOrigins = false,
 	__experimentalIsRenderedInSidebar = false,
-} ) {
+}: ColorPaletteProps ) {
```
Let's use the [`WordPressComponentProps` utility type](https://github.com/WordPress/gutenberg/blob/3c679aa7591f64e62305daeb8235f60bdcacca25/packages/components/src/ui/context/wordpress-component.ts), and refactor the component accordingly:

- the diff includes some duplicate changes to add a named export, as suggested in a previous comment from this round of review
- The `WordPressComponentProps` type adds support for polymorphism (ie. the `as` prop)
- The `WordPressComponentProps` type also includes types for all HTML attributes, including `className` — which therefore we can remove from the `ColorPaletteProps`
- We can forward all remaining `...otherProps` to the outer component (`VStack`, in this case)
- We can also add a forwarded ref in the process


```diff
diff --git a/packages/components/src/color-palette/index.tsx b/packages/components/src/color-palette/index.tsx
index a49c45e430..c777212759 100644
--- a/packages/components/src/color-palette/index.tsx
+++ b/packages/components/src/color-palette/index.tsx
@@ -1,6 +1,7 @@
 /**
  * External dependencies
  */
+import type { ForwardedRef } from 'react';
 import { colord, extend } from 'colord';
 import namesPlugin from 'colord/plugins/names';
 import a11yPlugin from 'colord/plugins/a11y';
@@ -9,7 +10,7 @@ import a11yPlugin from 'colord/plugins/a11y';
  * WordPress dependencies
  */
 import { __, sprintf } from '@wordpress/i18n';
-import { useCallback, useMemo } from '@wordpress/element';
+import { useCallback, useMemo, forwardRef } from '@wordpress/element';
 
 /**
  * Internal dependencies
@@ -30,6 +31,7 @@ import type {
 	PaletteObject,
 	SinglePaletteProps,
 } from './types';
+import type { WordPressComponentProps } from '../ui/context';
 
 extend( [ namesPlugin, a11yPlugin ] );
 
@@ -213,17 +215,21 @@ const areColorsMultiplePalette = (
 	);
 };
 
-export default function ColorPalette( {
-	clearable = true,
-	className,
-	colors = [],
-	disableCustomColors = false,
-	enableAlpha,
-	onChange,
-	value,
-	__experimentalHasMultipleOrigins = false,
-	__experimentalIsRenderedInSidebar = false,
-}: ColorPaletteProps ) {
+function UnforwardedColorPalette(
+	props: WordPressComponentProps< ColorPaletteProps, 'div' >,
+	forwardedRef: ForwardedRef< any >
+) {
+	const {
+		clearable = true,
+		colors = [],
+		disableCustomColors = false,
+		enableAlpha,
+		onChange,
+		value,
+		__experimentalHasMultipleOrigins = false,
+		__experimentalIsRenderedInSidebar = false,
+		...otherProps
+	} = props;
 	const clearColor = useCallback( () => onChange( undefined ), [ onChange ] );
 
 	const buttonLabelName = useMemo(
@@ -291,7 +297,7 @@ export default function ColorPalette( {
 	};
 
 	return (
-		<VStack spacing={ 3 } className={ className }>
+		<VStack spacing={ 3 } ref={ forwardedRef } { ...otherProps }>
 			{ ! disableCustomColors && (
 				<CustomColorPickerDropdown
 					isRenderedInSidebar={ __experimentalIsRenderedInSidebar }
@@ -350,3 +356,16 @@ export default function ColorPalette( {
 		</VStack>
 	);
 }
+
+/**
+ * TBD Needs a description of the component (which should be taken from the README )
+ *
+ * ```jsx
+ * import { ColorPalette } from '@wordpress/components';
+ *
+ * TBD code example
+ * ```
+ */
+export const ColorPalette = forwardRef( UnforwardedColorPalette );
+
+export default ColorPalette;
diff --git a/packages/components/src/color-palette/types.ts b/packages/components/src/color-palette/types.ts
index 1c8f6ec1a8..91e598faf1 100644
--- a/packages/components/src/color-palette/types.ts
+++ b/packages/components/src/color-palette/types.ts
@@ -46,10 +46,6 @@ export type ColorPaletteProps = {
 	 * Whether the palette should have a clearing button.
 	 */
 	clearable?: boolean;
-	/**
-	 * Classes to be applied to the container.
-	 */
-	className?: string;
 	/**
 	 * Array with the colors to be shown.
 	 *
```

Let me know if you have any questions on `WordPressComponentProps`

kienstra replied:
Thanks, committed in https://github.com/WordPress/gutenberg/pull/44632/commits/a79c187b64cec79c2026cf800dcaff7c2393d63b

ciampo reviewed (commented):
We're definitely making progress here! Added a few more comments to tidy up Storybook and types.

One thing that certainly needs improving is the component's README — we'll need to basically rewrite it so that it adheres to the latest format that we've been using, and including all the props information.

kienstra commented:
Thanks a lot, Marco! I'll apply your suggestions today.

kienstra committed:
Commit Marco's changes, including a named export

kienstra committed:
Add default tags, thanks to Marco's idea

kienstra committed:
Apply Marco's suggestion to remove ts-ignore

Add 'as CSSProperties'
to remove the need for ts-ignore

kienstra committed:
Apply Marco's suggestions, creating UnforwardedColorProps

The jsx example might not be right.
Also, I added a description for the component
in the JS DocBlock, as there wasn't one in README.md.
But maybe that's not right.

kienstra commented on packages/components/src/color-palette/index.tsx:
```diff
@@ -284,22 +342,45 @@ export default function ColorPalette( {
 }
+
+/**
+ * Allows selecting from given colors.
```
Maybe this JSDoc block isn't what you [had in mind](https://github.com/WordPress/gutenberg/pull/44632#discussion_r999310616). 

I didn't see a description in [README.md](https://github.com/WordPress/gutenberg/blob/5a82cf5723b0feb0de9ec23e267f6b190d2a5ad0/packages/components/src/color-palette/README.md), so I added this, which might be completely off 😄

ciampo replied:
It looks good for now, we can always refine it before merging — thank you!

kienstra replied:
Thanks, Marco!

kienstra commented on packages/components/src/color-palette/index.tsx:
````diff
@@ -284,22 +342,45 @@ export default function ColorPalette( {
+ * Allows selecting from given colors.
+ *
+ * ```jsx
+ * import { ColorPalette } from '@wordpress/components';
````
Copy-pasted verbatim from [README.md](https://github.com/WordPress/gutenberg/blob/5a82cf5723b0feb0de9ec23e267f6b190d2a5ad0/packages/components/src/color-palette/README.md#usage).

Maybe that's not what you were thinking, or maybe we could use more examples.

ciampo replied:
Same as above — it's definitely good enough for now, we can always polish it later

kienstra replied:
Thanks!

kienstra committed:
Merge in trunk, resolve conflict

The only conflict was trivial, in
components/CHANGELOG.md

kienstra committed:
Fix a linting error, remove needless className

kienstra commented:
Hi @ciampo,
Thanks a lot for your suggestions, they're applied now.

>One thing that certainly needs improving is the component's README — we'll need to basically rewrite it so that it adheres to the latest format that we've been using, and including all the props information.

Would you like to suggest the changes in `README`, or should I start those?

Thanks, Marco!

Update: sorry for spelling your name Marcos before, a friend is named Marcos 😄

ciampo commented:
> Would you like to suggest the changes in README, or should I start those?

Let's leave the README as the last thing, in case we make further adjustments to the component in the meantime.

As the next steps, could you refactor `styles.js` and `test/index.js` to TypeScript too?

The unit tests in particular are a great way to double check that our tests are actually using the component by respecting the types that we've just added in this PR.

ciampo commented on packages/components/src/color-palette/stories/index.tsx:
```diff
@@ -6,34 +12,24 @@ import { useState } from '@wordpress/element';
-				type: 'boolean',
-			},
-		},
+		onChange: { action: 'onChange', control: { type: null } },
```
Let's also disable controls for the `as` prop:
```suggestion
		as: { control: { type: null } },
		onChange: { action: 'onChange', control: { type: null } },
```

kienstra replied:
Thanks, committed in https://github.com/WordPress/gutenberg/pull/44632/commits/8001f145d7633d013b27d369e80d83bdc89b2864

ciampo commented on packages/components/src/color-palette/types.ts:
```diff
@@ -0,0 +1,89 @@
+	 */
+	colors?: ( PaletteObject | ColorObject )[];
+	/**
+	 * Whether to allow custom color or not.
```
Let's clarify this prop description

```suggestion
	 * Whether to allow the user to pick a custom color on top of the predefined
	 * choices (defined via the `colors` prop).
```

kienstra replied:
Committed in https://github.com/WordPress/gutenberg/pull/44632/commits/e793921d09c7bafdbf90d170efcfdf28630990bc

kienstra committed:
Commit Marco's suggestion: Update packages/components/src/color-palette/stories/index.tsx

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com>

kienstra committed:
Commit Marco's suggestion: Update packages/components/src/color-palette/types.ts

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com>

kienstra committed:
Rename test/index.js to test/indes.tsx, mv snapshot

kienstra committed:
Add types to test/index.tsx

kienstra committed:
Renamce styles.js to styles.ts

kienstra commented:
Hi @ciampo,
>As the next steps, could you refactor styles.js and test/index.js to TypeScript too?

Sure, I just did that.

* [styles.js](https://github.com/WordPress/gutenberg/blob/430fb9faadb00a5be74e414a7975bec672a3bba4/packages/components/src/color-palette/styles.js ) is renamed to `.ts`, with no change
* [test__snapshots__/index.js.snap](https://github.com/WordPress/gutenberg/blob/430fb9faadb00a5be74e414a7975bec672a3bba4/packages/components/src/color-palette/test/__snapshots__/index.js.snap) is renamed to `.tsx`, with no change
* [test/index.js](https://github.com/WordPress/gutenberg/blob/430fb9faadb00a5be74e414a7975bec672a3bba4/packages/components/src/color-palette/test/index.js) is renamed to [.tsx](https://github.com/WordPress/gutenberg/blob/293fafe9564f54dfbaa8b101eda29032e92e9e2d/packages/components/src/color-palette/test/index.tsx), with only a [few changes](https://github.com/WordPress/gutenberg/pull/44632/files#diff-0e9f6fa7c3d085a5471e4eac0d4afc563cb68f6487fbeaa208dc0f117077048b)

Maybe I missed some changes, though.

Thanks for your huge amount of patience!

ciampo commented on packages/components/src/color-palette/types.ts:
```diff
@@ -0,0 +1,90 @@
+	 */
+	clearable?: boolean;
+	/**
+	 * Array with the colors to be shown.
```
```suggestion
	 * Array with the colors to be shown. When displaying multiple color palettes
	 * to choose from, the format of the array changes from an array of colors
	 * objects, to an array of color palettes.
```

ciampo commented on packages/components/src/color-palette/test/index.tsx:
```diff
@@ -7,16 +7,17 @@ import userEvent from '@testing-library/user-event';
 describe( 'ColorPalette', () => {
-	const colors = [
+	const colors: ColorObject[] = [
 		{ name: 'red', color: '#f00' },
 		{ name: 'white', color: '#fff' },
 		{ name: 'blue', color: '#00f' },
 	];
 	const currentColor = '#f00';
-	const onChange = jest.fn();
+	const onChange: jest.Mock< OnColorChange > = jest.fn();
```
There shouldn't be a need for those changes — have you tried not adding those extra types?

If we can avoid the types, then I'd argue that we also don't need to export the `OnColorChange` type from `types.ts` either.

kienstra replied:
Good idea, I don't think those are needed either. I'll test it and revert them.

kienstra replied:
Reverted in https://github.com/WordPress/gutenberg/pull/44632/commits/d77ca5b7f913e10dcc9e0607382ce0681dfe2666

ciampo reviewed (commented):
> Thanks for your huge amount of patience!

Likewise!

Overall, I think the code is in a pretty good state.

The last step is to update the README file — you can find more info (and an example) in [the contributing guidelines](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/CONTRIBUTING.md#documentation):

- We will need to provide a generic component description — we can probably go with something like "`ColorPalette` allows the user to pick a color from a list of pre-defined color entries" — What do you think?
- It looks like the current README already has a `Usage` section, which can be reused
- For the list of props, let's use the props defined in `types.ts` (and their descriptions) as the source of truth

Finally, we should update [the JSDoc for the default export](https://github.com/WordPress/gutenberg/blob/a79c187b64cec79c2026cf800dcaff7c2393d63b/packages/components/src/color-palette/index.tsx#L360-L383) with whatever description and usage code snippet were used in the README.

Once we're done with those changes, I believe that we can have a final round of review!

kienstra committed:
Commit Marco's suggestion: Update packages/components/src/color-palette/types.ts

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com>

kienstra committed:
Revert "Add types to test/index.tsx"

This reverts commit 06f7c4edc739ff51a07749db6f5571a88cf94e44.

kienstra committed:
Paste Marco's description verbatim

https://github.com/WordPress/gutenberg/pull/44632#pullrequestreview-1153120315

kienstra committed:
Copy props verbatim from types.ts into README.md

kienstra committed:
Update the JSDoc description to be the same as the README.md description

The usage example was already the same
as in the README.md

kienstra commented:
>We will need to provide a generic component description — we can probably go with something like "ColorPalette allows the user to pick a color from a list of pre-defined color entries" — What do you think?

That's a good description, I [committed it verbatim](https://github.com/WordPress/gutenberg/pull/44632/commits/626b25d6babda6f03115903cc759254acb939184). It's not too long, and lets you know what it does.

>For the list of props, let's use the props defined in types.ts (and their descriptions) as the source of truth

Thanks, I [pasted the props from types.ts](https://github.com/WordPress/gutenberg/pull/44632/commits/905c40e29544059746b109aac4d5cb74a2c95f91).


>Finally, we should update [the JSDoc for the default export](https://github.com/WordPress/gutenberg/blob/a79c187b64cec79c2026cf800dcaff7c2393d63b/packages/components/src/color-palette/index.tsx#L360-L383) with whatever description and usage code snippet were used in the README.

Sounds good, I [updated the description](https://github.com/WordPress/gutenberg/pull/44632/commits/eee647c93db0a7faf6777eb1bd54766b415c96ab). The usage example was already the same as in `README.md`.

kienstra committed:
Merge branch 'trunk' into update/color-palette-component-ts

kienstra committed:
Remove extra entry for Tooltip

I think I added this
when merging in trunk and resolving the conflicts.

kienstra committed:
Move the CHANGELOG entry up, to Unreleased

kienstra commented:
Hi @ciampo,
The commits above address your points 😄 

>We will need to provide a generic component description — we can probably go with something like "ColorPalette allows the user to pick a color from a list of pre-defined color entries" — What do you think?

That's a good description, I [committed it verbatim](https://github.com/WordPress/gutenberg/pull/44632/commits/626b25d6babda6f03115903cc759254acb939184). It's not too long, and lets you know what it does.

>For the list of props, let's use the props defined in types.ts (and their descriptions) as the source of truth

Thanks, I [pasted the props from types.ts](https://github.com/WordPress/gutenberg/pull/44632/commits/905c40e29544059746b109aac4d5cb74a2c95f91).


>Finally, we should update [the JSDoc for the default export](https://github.com/WordPress/gutenberg/blob/a79c187b64cec79c2026cf800dcaff7c2393d63b/packages/components/src/color-palette/index.tsx#L360-L383) with whatever description and usage code snippet were used in the README.

Sounds good, I [updated the description](https://github.com/WordPress/gutenberg/pull/44632/commits/eee647c93db0a7faf6777eb1bd54766b415c96ab). The usage example was already the same as in `README.md`.

ciampo commented on packages/components/src/color-palette/README.md:
```diff
@@ -44,12 +54,28 @@ classes to be applied to the container.
+-   Required: No
+-   Default: false
+
 ## Usage
```
As suggested in the [guidelines](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/CONTRIBUTING.md#documentation), let's move the `Usage` section at the top of the document

```diff
diff --git a/packages/components/src/color-palette/README.md b/packages/components/src/color-palette/README.md
index 5d3ed4601c..49612e3f91 100644
--- a/packages/components/src/color-palette/README.md
+++ b/packages/components/src/color-palette/README.md
@@ -2,6 +2,36 @@
 
 `ColorPalette` allows the user to pick a color from a list of pre-defined color entries.
 
+## Usage
+
+```jsx
+import { ColorPalette } from '@wordpress/components';
+import { useState } from '@wordpress/element';
+
+const MyColorPalette = () => {
+	const [ color, setColor ] = useState ( '#f00' )
+	const colors = [
+		{ name: 'red', color: '#f00' },
+		{ name: 'white', color: '#fff' },
+		{ name: 'blue', color: '#00f' },
+	];
+
+	return (
+		<ColorPalette
+			colors={ colors }
+			value={ color }
+			onChange={ ( color ) => setColor( color ) }
+		/>
+	);
+} );
+```
+
+If you're using this component outside the editor, you can
+[ensure `Tooltip` positioning](/packages/components/README.md#popovers-and-tooltips)
+for the `ColorPalette`'s color swatches, by rendering your `ColorPalette` with a
+`Popover.Slot` further up the element tree and within a
+`SlotFillProvider` overall.
+
 ## Props
 
 The component accepts the following props.
@@ -75,33 +105,3 @@ Whether this is rendered in the sidebar.
 -   Type: `Boolean`
 -   Required: No
 -   Default: false
-
-## Usage
-
-```jsx
-import { ColorPalette } from '@wordpress/components';
-import { useState } from '@wordpress/element';
-
-const MyColorPalette = () => {
-	const [ color, setColor ] = useState ( '#f00' )
-	const colors = [
-		{ name: 'red', color: '#f00' },
-		{ name: 'white', color: '#fff' },
-		{ name: 'blue', color: '#00f' },
-	];
-
-	return (
-		<ColorPalette
-			colors={ colors }
-			value={ color }
-			onChange={ ( color ) => setColor( color ) }
-		/>
-	);
-} );
-```
-
-If you're using this component outside the editor, you can
-[ensure `Tooltip` positioning](/packages/components/README.md#popovers-and-tooltips)
-for the `ColorPalette`'s color swatches, by rendering your `ColorPalette` with a
-`Popover.Slot` further up the element tree and within a
-`SlotFillProvider` overall.
```

kienstra replied:
Thanks, committed in https://github.com/WordPress/gutenberg/pull/44632/commits/4fd5d9c8b1fc2af1e70ce66c02506ed426463da3

ciampo commented on packages/components/src/color-palette/README.md:
```diff
@@ -1,21 +1,31 @@
 # ColorPalette
 
+ColorPalette allows the user to pick a color from a list of pre-defined color entries.
```
```suggestion
`ColorPalette` allows the user to pick a color from a list of pre-defined color entries.
```

kienstra replied:
Thanks, committed in https://github.com/WordPress/gutenberg/pull/44632/commits/d65a832ab8b85ab4f993eb3c2ad3b4c098ad5ead

ciampo commented on packages/components/src/color-palette/README.md:
```diff
@@ -1,21 +1,31 @@
 The component accepts the following props.
 
-{ colors, disableCustomColors = false, value, onChange, className, clearable = true }
+{ colors = [],  disableCustomColors = false, enableAlpha, value, onchange, className, clearable = true, __experimentalHasMultipleOrigins = false, __experimentalIsRenderedInSidebar = false }
```
```diff
diff --git a/packages/components/src/color-palette/README.md b/packages/components/src/color-palette/README.md
index 49612e3f91..b194f3cef0 100644
--- a/packages/components/src/color-palette/README.md
+++ b/packages/components/src/color-palette/README.md
@@ -36,8 +36,6 @@ for the `ColorPalette`'s color swatches, by rendering your `ColorPalette` with a
 
 The component accepts the following props.
 
-{ colors = [],  disableCustomColors = false, enableAlpha, value, onchange, className, clearable = true, __experimentalHasMultipleOrigins = false, __experimentalIsRenderedInSidebar = false }
-
 ### colors
 Array with the colors to be shown. When displaying multiple color palettes to choose from, the format of the array changes from an array of colors objects, to an array of color palettes.
```

kienstra replied:
Committed in https://github.com/WordPress/gutenberg/pull/44632/commits/29984c0cd55611a12d80b6e2d2cedb5debb2b8bc

ciampo commented on packages/components/src/color-palette/README.md:
```diff
@@ -1,21 +1,31 @@
+{ colors = [],  disableCustomColors = false, enableAlpha, value, onchange, className, clearable = true, __experimentalHasMultipleOrigins = false, __experimentalIsRenderedInSidebar = false }
 
 ### colors
-
-Array with the colors to be shown.
+Array with the colors to be shown. When displaying multiple color palettes to choose from, the format of the array changes from an array of colors objects, to an array of color palettes.
 
 -   Type: `Array`
--   Required: Yes
+-   Required: No
+-   Default: `[]`
```
A bit nitty, but let's try to follow the format highlighted in the [guidelines](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/CONTRIBUTING.md#readme-example):

```markdown
### `propName`: Typescript style type i.e `string`, `number`, `( nextValue: string ) => void`

Prop description. With a new line before and after the description and before and after type/required blocks.

-   Required: Either `Yes` or `No`
<!-- If the prop has a default value, add the following line: -->
-   Default: [default value]
```

For example, for the `color` prop the docs would change to

```markdown
### `colors`: `( PaletteObject | ColorObject )[]`

Array with the colors to be shown. When displaying multiple color palettes to choose from, the format of the array changes from an array of colors objects, to an array of color palettes.

-   Required: No
-   Default: `[]`
```

kienstra replied:
Oh, good idea. Sorry for missing that in CONTRIBUTING.md. 

Committed in https://github.com/WordPress/gutenberg/pull/44632/commits/988e9401da364236c2dfe0296fe4cd234029ae4b

ciampo commented on packages/components/src/color-palette/README.md:
```diff
@@ -44,12 +54,28 @@ classes to be applied to the container.
 -   Required: No
 -   Default: true
 
+### __experimentalHasMultipleOrigins
+
+Whether the colors prop is an array of color palettes, rather than an array of color objects.
+
+-   Type: `Boolean`
+-   Required: No
+-   Default: false
+
+### __experimentalIsRenderedInSidebar
+
+Whether this is rendered in the sidebar.
+
+-   Type: `Boolean`
+-   Required: No
+-   Default: false
+
```
Let's actually not list experimental props in the README — apologies if I forgot to mention that earlier!

kienstra replied:
Good idea, removed in https://github.com/WordPress/gutenberg/pull/44632/commits/4fd5d9c8b1fc2af1e70ce66c02506ed426463da3

ciampo reviewed (commented):
Just a few tweaks to the README!

kienstra commented:
Thanks, good ideas! I'll apply these today.

kienstra committed:
Move Usage section to the top of the README. remove experimental props from README

kienstra committed:
Commit Marco's suggestion: Update packages/components/src/color-palette/README.md

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com>

kienstra committed:
Remove the example of the full props

kienstra committed:
Change the props format to match CONTRIBUTING.md

https://github.com/WordPress/gutenberg/blob/a42805e157f6c6933f4ef7cabcfc87fa3af81aea/packages/components/CONTRIBUTING.md#readme-example

kienstra committed:
Merge in trunk, resolve conflict

Resolve trivial conflict
in CHANGELOG.md.

kienstra commented:
Hi @ciampo,
Thanks a lot for your patience. The commits above apply your suggestions.

ciampo commented on packages/components/src/color-palette/stories/index.tsx:
```diff
@@ -42,14 +39,26 @@ const meta = {
+					onChange?.( newColor );
+				} }
+			/>
+			{ /* @ts-ignore Property 'Slot' does not exist on type. */ }
```
Let's use `@ts-expect-error` instead and add a more specific reason

```suggestion
			{ /* @ts-expect-error The 'Slot' component hasn't been typed yet. */ }
```

ciampo commented on packages/components/src/color-palette/types.ts:
```diff
@@ -0,0 +1,92 @@
+	colors: PaletteObject[];
+};
+
+export type CustomColorPickerDropdownProps = {
```
```suggestion
// TODO: should extend `Dropdown`'s props once it gets refactored to TypeScript
export type CustomColorPickerDropdownProps = {
```

ciampo commented on packages/components/src/color-palette/types.ts:
```diff
@@ -0,0 +1,92 @@
+
+export type CustomColorPickerDropdownProps = {
+	isRenderedInSidebar: boolean;
+	renderContent: () => JSX.Element;
```
Let's use `ReactNode` as the return type

```suggestion
	renderContent: () => ReactNode;
```

ciampo commented on packages/components/src/color-palette/types.ts:
```diff
@@ -0,0 +1,92 @@
+	renderToggle: ( props: {
+		isOpen: boolean;
+		onToggle: MouseEventHandler< HTMLButtonElement >;
+	} ) => JSX.Element;
```
Same as above:

```suggestion
	} ) => ReactNode;
```

ciampo reviewed (approved):
Left a few minor comments, but once they are addressed we can merge this PR!

 - ✅ Code changes LGTM
 - ✅ Storybook examples work as expected
 - ✅ No regressions spotted in the editor

kienstra committed:
Commit Marco's suggestion: Update packages/components/src/color-palette/stories/index.tsx

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com>

kienstra committed:
Commit Marco's suggestion: Update packages/components/src/color-palette/types.ts

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com>

kienstra committed:
Commit Marco's suggestion: Update packages/components/src/color-palette/types.ts

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com>

kienstra committed:
Commit Marco's suggestion: Update packages/components/src/color-palette/types.ts

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com>

kienstra commented:
Thanks, Marco! Just committed your suggestions.

ciampo committed:
Merge branch 'trunk' into update/color-palette-component-ts

kienstra commented:
Hi Marco, thanks for your great ideas on the types and documentation here! And thanks for your patience.

ciampo commented:
> Hi Marco, thanks for your great ideas on the types and documentation here! And thanks for your patience.

Thank you!

If you enjoyed this type of work, we have more components in need of a TypeScript refactor #35744 

I'm sure that the next refactors will be quicker, as you get used to the migration requirements :)

kienstra commented:
Thanks, Marco! I'll open another PR to convert another component for https://github.com/WordPress/gutenberg/issues/35744.
